storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,756 @@
+            blobs.insert(index, blob);
+        }
+        if !blobs.is_empty() {
+            // Check that there are no gaps in the blob numbering, which would indicate missing data.
+            let mut it = blobs.iter();
+            let mut previous_index = *it.next().unwrap().0;
+            for (index, _) in it {
+                if *index != previous_index + 1 {
+                    return Err(Error::MissingBlob(previous_index + 1));
+                }
+                previous_index = *index;
+            }
+        } else {
+            debug!("no blobs found");
+            let blob = runtime.open(&cfg.partition, &0u64.to_be_bytes()).await?;
+            blobs.insert(0, blob);
+        }
+
+        // Initialize metrics
+        let tracked = Gauge::default();
+        let synced = Counter::default();
+        let pruned = Counter::default();
+        {
+            let mut registry = cfg.registry.lock().unwrap();
+            registry.register("tracked", "Number of blobs", tracked.clone());
+            registry.register("synced", "Number of syncs", synced.clone());
+            registry.register("pruned", "Number of blobs pruned", pruned.clone());
+        }
+        tracked.set(blobs.len() as i64);
+
+        // truncate the last blob if it's not the expected length, which might happen from unclean shutdown.
+        let newest_blob = blobs.last_key_value().unwrap().1;
+        let blob_len: u64 = newest_blob.len().await?;
+        let chunk_size: usize = size_of::<u32>() + N;
+        if blob_len % chunk_size as u64 != 0 {
+            warn!(
+                "last blob len ({}) is not a multiple of item size, truncating",
+                blob_len
+            );
+            newest_blob
+                .truncate(blob_len - blob_len % chunk_size as u64)
+                .await?;
+            newest_blob.sync().await?;
+        }
+
+        Ok(Self {
+            runtime,
+            cfg,
+            blobs,
+            tracked,
+            synced,
+            pruned,
+        })
+    }
+
+    /// Sync any pending updates to disk.
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        self.synced.inc();
+        let newest_blob = self.newest_blob();
+        debug!("syncing blob {}", newest_blob.0);
+        self.newest_blob().1.sync().await.map_err(Error::Runtime)
+    }
+
+    /// Returns the total number of items in the journal, ignoring any pruning. The next value
+    /// appended to the journal will be at this position.
+    pub async fn size(&self) -> Result<u64, Error> {
+        let chunk_size = size_of::<u32>() + N;
+        let newest_blob = self.newest_blob();
+        let blob_len = newest_blob.1.len().await?;
+        assert_eq!(blob_len % chunk_size as u64, 0);
+        let items_in_blob = blob_len / chunk_size as u64;
+        Ok(items_in_blob + self.cfg.items_per_blob * newest_blob.0)
+    }
+
+    /// Append a new item to the journal. Return the item's position in the journal, or error if the
```

Comment:
nit: add a small note that this is the "itemth" index not the byte position

Reply:
I took the approach of providing a definition of `position` in the file doc header. Would you still like me to clarify it here?